### ISSUE TITLE
Build error with ghc-8.10.1

### ISSUE THREAD

**Issue by peti** [rblaze/haskell-dbus#41]
I have tried to compile `dbus-1.2.11` with the release candidate for `ghc-8.10.1` and got the following error:

~~~
[  157s] lib/DBus/Generation.hs:235:79: error:
[  157s]     • Couldn't match type ‘Exp’ with ‘Maybe Exp’
[  157s]       Expected type: [Exp] -> Exp
[  157s]         Actual type: [Maybe Exp] -> Exp
[  157s]     • In the third argument of ‘mapOrHead'’, namely ‘TupE’
[  157s]       In the expression:
[  157s]         mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
[  157s]       In an equation for ‘fromVariantExp’:
[  157s]           fromVariantExp
[  157s]             = mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
[  157s]     |
[  157s] 235 |         fromVariantExp = mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
[  157s]     |                                                                               ^^^^
[  157s] 
[  157s] lib/DBus/Generation.hs:236:61: error:
[  157s]     • Couldn't match type ‘Exp’ with ‘Maybe Exp’
[  157s]       Expected type: [Exp] -> Exp
[  157s]         Actual type: [Maybe Exp] -> Exp
[  157s]     • In the third argument of ‘mapOrHead'’, namely ‘TupE’
[  157s]       In the expression: mapOrHead' VarE finalOutputNames TupE
[  157s]       In an equation for ‘finalResultTuple’:
[  157s]           finalResultTuple = mapOrHead' VarE finalOutputNames TupE
[  157s]     |
[  157s] 236 |         finalResultTuple = mapOrHead' VarE finalOutputNames TupE
[  157s]     |                                                             ^^^^
[  157s] 
[  157s] lib/DBus/Generation.hs:435:79: error:
[  157s]     • Couldn't match type ‘Exp’ with ‘Maybe Exp’
[  157s]       Expected type: [Exp] -> Exp
[  157s]         Actual type: [Maybe Exp] -> Exp
[  157s]     • In the third argument of ‘mapOrHead'’, namely ‘TupE’
[  157s]       In the expression:
[  157s]         mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
[  157s]       In an equation for ‘fromVariantExp’:
[  157s]           fromVariantExp
[  157s]             = mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
[  157s]     |
[  157s] 435 |         fromVariantExp = mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
[  157s]     |                                                                               ^^
~~~

Is this, by any chance, easy to fix?

**Comment by rblaze** [comment]
Looks like the type of `TupE` was changed in `template-haskell`. It could be easy to fix for 8.10, but I'm not sure how to do this w/o breaking the same code for 8.8 and previous versions. I wonder how many other packages will be broken.

Which version of `template-haskell` comes with 8.10? The latest on hackage is 2.15, and it still has correct `TupE`.

**Comment by mimi1vx** [comment]
@rblaze maybe this as inspiration? https://github.com/ekmett/lens/commit/12768f14b1112d4f4f1036861197e0610c325853#diff-ccec92295146aff0f48c65899fd6ad05

more in https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#standard-cpp-macros

> Which version of template-haskell comes with 8.10? The latest on hackage is 2.15, and it still has correct TupE.

2.16.0.0 and from ghc release notes:
https://ghc.gitlab.haskell.org/ghc/doc/users_guide/8.10.1-notes.html#template-haskell

**Comment by peti** [comment]
> Which version of `template-haskell` comes with 8.10?

`template-haskell-2.16.0.0`

**Comment by rblaze** [comment]
I'll wait for the new version to hit hackage, to pick up all the changes in one go.

**Comment by peti** [comment]
Ping? Will there be a fix?

**Comment by rblaze** [comment]
The fix builds on my machine, but something fails in Travis builds, so I can't verify it actually works: https://travis-ci.org/github/rblaze/haskell-dbus/jobs/668408618

I'll get back to it a bit later.

**Comment by aavogt** [comment]
Another way to fix this would be to use `tupE :: [ExpQ] -> ExpQ` which was not changed. But on the other hand, using ExpQ means changing more code: https://github.com/aavogt/haskell-dbus/commit/42879431713c937670a22a055ef4017d6f9f389d

**Comment by felixonmars** [comment]
Hopefully addressed by #48

**Comment by colonelpanic8** [comment]
@rblaze you can ping me for fixes related to the generation template haskell code, since I'm pretty sure that I'm the only one that has touched that.

**Comment by rblaze** [comment]
Published 1.2.14
@felixonmars thanks for the diff!
